lib/grandpa/round_test.go

```diff
@@ -216,388 +218,775 @@ func TestGrandpa_DifferentChains(t *testing.T) {
+			},
+			defineBlockTree: func(t *testing.T, blockState BlockState, _ []*Service) {
+				const withBranches = false
+				const baseLenght = 4
+				state.AddBlocksToState(t, blockState.(*state.BlockState), baseLenght, withBranches)
+			},
+		},
+
+		"varying_chain": {
+			voters: []*ed25519.Keypair{
+				ed25519Keyring.Alice().(*ed25519.Keypair),
+				ed25519Keyring.Bob().(*ed25519.Keypair),
+				ed25519Keyring.Charlie().(*ed25519.Keypair),
+				ed25519Keyring.Ian().(*ed25519.Keypair),
+			},
+			defineBlockTree: func(t *testing.T, blockState BlockState, neighbourServices []*Service) {
+				const diff uint = 5
+				rand := uint(rand.Intn(int(diff)))
+
+				const withBranches = false
+				const baseLenght = 4
+				headers, _ := state.AddBlocksToState(t, blockState.(*state.BlockState),
+					baseLenght+rand, withBranches)
+
+				// sync the created blocks with the neighbour services
+				// letting them know about those blocks
+				for _, ns := range neighbourServices {
+					for _, header := range headers {
+						block := &types.Block{
+							Header: *header,
+							Body:   types.Body{},
+						}
+						ns.blockState.(*state.BlockState).AddBlock(block)
+					}
+				}
+			},
+		},
+
+		"with_equivocations": {
+			voters: []*ed25519.Keypair{
+				ed25519Keyring.Alice().(*ed25519.Keypair),
+				ed25519Keyring.Bob().(*ed25519.Keypair),
+				ed25519Keyring.Charlie().(*ed25519.Keypair),
+				ed25519Keyring.Dave().(*ed25519.Keypair),
+				ed25519Keyring.Ian().(*ed25519.Keypair),
+			},
+			// alice and charlie equivocates
+			// it is a map as it is easy to check
+			whoEquivocates: map[int]struct{}{
+				3: {},
+				4: {},
+			},
+			defineBlockTree: func(t *testing.T, blockState BlockState, _ []*Service) {
+				// this creates a tree with 2 branches starting at depth 2
+				branches := map[uint]int{2: 1}
+				const baseLenght = 4
+				state.AddBlocksToStateWithFixedBranches(t, blockState.(*state.BlockState), baseLenght, branches)
+			},
+		},
+	}
+
+	for tname, tt := range tests {
+		tt := tt
+		t.Run(tname, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			grandpaServices := make([]*Service, len(tt.voters))
+			grandpaVoters := make([]types.GrandpaVoter, 0, len(tt.voters))
+
+			for _, kp := range tt.voters {
+				grandpaVoters = append(grandpaVoters, types.GrandpaVoter{
+					Key: *kp.Public().(*ed25519.PublicKey),
+				})
 			}
-		}
-	}
-}
 
-func cleanup(gs *Service, in chan *networkVoteMessage, done *bool) {
-	*done = true
-	close(in)
-	gs.cancel()
-}
+			for idx := range tt.voters {
+				// gossamer gossip a prevote/precommit message and then waits `subroundInterval` * 4
+				// to issue another prevote/precommit message
+				const subroundInterval = 100 * time.Millisecond
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
+
+				st := newTestState(t)
+				grandpaServices[idx] = &Service{
+					ctx:          ctx,
+					cancel:       cancel,
+					paused:       atomic.Value{},
```

Comment:
I don't think that's needed

```suggestion
```

Reply:
This is important otherwise we will face a panic exception as we load the value stored in it and if we don't create or store a value (see the line right after the creation of the service) in it then it will panic